### ISSUE TITLE
Translation of code to C++ or C

### ISSUE THREAD

**Issue by mambastudio** [madmann91/hagrid#2]
Hi,

Not really an issue, but was curious if there are any plans in the future to convert the code to C++ or C for the people who want to experiment in other languages or to understand in detail the algorithms in the code? It will be an amazing project if there was such an initiative. 

Anyway, great concept, and seems this is the ideal first of a kind concept for solving the teapot in a stadium problem.

Joe.

**Comment by madmann91** [comment]
I am not sure I understand the question correctly. The code is written in C++ and CUDA. If you mean converting the code to pure C (still using CUDA), I don't think I would have the time to do that, since it is quite an undertaking. That said, the code does not use C++ features too heavily, only templates are used here and there. The code itself is mostly procedural, and writing C bindings shouldn't be too hard if you want to integrate it into some other project. There's also a dependency on the C++ library CUB for high-performance device-wide parallel primitives, which may be an issue if you want to port the code to pure C.

If you mean removing the dependency on CUDA, then this is even harder since everything is designed to run on GPUs. I don't think that would make much sense to use this code as a starting point for a CPU implementation, as it would be probably easier to start from scratch. Instead, there is a C++ implementation of the construction algorithm available in the digital EG lib: https://onlinelibrary.wiley.com/action/downloadSupplement?doi=10.1111%2Fcgf.13142&file=cgf13142-sup-0001-S1.zip. This implementation still uses OpenCL to run the traversal kernel on the GPU, and it does not have all the features this repository offers, but it can build the data structure entirely on the CPU in parallel using TBB.

**Comment by madmann91** [comment]
Also, if you're looking for a C++ CPU ray-tracing library, you may be interested in checking this out: https://github.com/madmann91/bvh (sorry for the shameless self-promotion).

**Comment by mambastudio** [comment]
@madmann91 

Oh, thank you for the very detail reply, and quite an honor. 

First of all, I'm a huge fan of the https://github.com/madmann91/bvh (bvh project), which I came across it like a month ago, and I've been going through the code in detail. Great project, and it came as a blessing in understanding various algorithms for the bvh. I was inspired particularly by Ploc which [I implemented](https://github.com/mambastudio/MambaTracer/blob/master/src/cl/core/kernel/Ploc.cl) with OpenCL code in [my java project](https://github.com/mambastudio/MambaTracer), and damn it's fast in construction and thus future direction will be to investigate parallel reinsertion optimizer, since the ploc bvh tree is not quite high quality even if you use larger radius on complex scenes with sacrifice on speed (sam miguel for example). And my ray tracer is stackless, hence bitstack for deeper tree traversal is an avenue for it to crush if there is an overflow. So I really appreciate you coming up with the project.

I was actually thinking along those lines of the bvh project in terms of removing the dependencies on CUDA, since I'm not well accustomed to CUDA (specifically on certain algorithms like emit_new_refs which you use CUDA intrinsic codes such as __ballot_sync, __shfl_sync and some bitwise manipulations), hence porting becomes difficult without understanding what's happening in the code. I think the idea is to understand the concepts of approach in code on certain functions as described in the main paper, and eventually becomes an avenue to port to other code (that's how I did with my Ploc code). I can start with the C++ code you've provided in the link, as a starting point. 

I really appreciate a lot.
